src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -120,8 +120,14 @@ private static JsonNode validateAndParseSemiStructuredAsJsonTree(
    * @return JSON string representing the input
    */
   static String validateAndParseVariant(String columnName, Object input) {
-    String output =
-        validateAndParseSemiStructuredAsJsonTree(columnName, input, "VARIANT").toString();
+    JsonNode node = validateAndParseSemiStructuredAsJsonTree(columnName, input, "VARIANT");
+
+    // Missing nodes are not valid json, ingest them as NULL instead
+    if (node.isMissingNode()) {
+      return null;
+    }
```

Comment:
This will not work because null is not an ordinary value, it needs some special treatment in both arrow and parquet row buffers. We need to increase null count, check if the column is nullable, etc. 

For Arrow, take a look at `net.snowflake.ingest.streaming.internal.ArrowRowBuffer#insertNull`. For Parquet it is the branch at `net/snowflake/ingest/streaming/internal/ParquetValueParser.java:127`. This code needs to be triggered when we ingest null.

Reply:
Until now, `NULL` was stored in the data file only if user directly ingested java `null`. Now, the user can insert non-null value (an empty string in this case) and we would like to ingest null. The code for will need to be restructured a bit to make sure we still trigger the same code path as if the user ingested null directly.

Reply:
Thanks for the context and detailed explanation, this is very helpful, let me investigate further and update the PR

Reply:
I changed code structure to handle returning null from the validation by doing the null check after getting a parsedValue object. 

Another approach that I could think about was changing the return type of validateAndParseXXXX() methods to return Optional<String> to force handle the null cases in the downstream calls in a more defensive manner, i am not sure about the convention here so I'll leave that up to you to decide if returning null is okay here

Reply:
I think returning null from the validate() methods is fine.